File: day5/day5.py
#!/usr/bin/python

# wrong answer ...

import math

def match(a, b):
    matches = (abs(ord(a) - ord(b)) == 32)
    # print("{} {} {}".format(a, b, matches))
    return matches

def compact(txt):
    i = 0
    while i < (len(txt)-1):
        if match(txt[i], txt[i+1]):
            txt = txt[:i] + txt[i+2:]
        else:
            i += 1
    return txt

if __name__ == "__main__":

    with open("input_0.txt") as f:
        txt = f.read()

    change = True
    while change:
        l = len(txt)
        txt = compact(txt)   
        change = (len(txt) != l)

    print(len(txt))
    print (txt)

File: day5/Day5.hs
import Data.Char

match :: Char -> Char -> Bool
match a aa = (abs $ (ord a) - (ord aa)) == 32

myZip :: [Char] -> [Char]
myZip xs = myZip' [] xs
  where
    myZip' ys [] = ys
    myZip' [] (x:xs) = myZip' [x] xs 
    myZip' ys (x:xs)
      | match (last ys) x = myZip' (init ys) xs
      | otherwise = myZip' (ys ++ [x]) xs

-- after python giving the same wrong answer, I took a correct algorithm from reddit
part1 :: String -> Int
part1 = length . foldr step ""
  where
    step x (y:ys) | x /= y && toUpper x == toUpper y = ys
    step x ys                                        = x : ys


main = do
  raw <- readFile "input_0.txt"
  print $ length raw
  let res = myZip raw
  print $ length res
  let res = myZip raw
  print $ length res
  let res = myZip raw
  print $ length res
